IntAct.pm

```diff
@@ -0,0 +1,380 @@
+    }
+  }
+   
+  return \%header;
+}
+
+# match lines of IntAct data file using HGVS id
+sub _match_id {
+  my ($self, $id_ref, $id_des) = @_;
+
+  my $param_hash = $self->params_to_hash();
+  my $mutation_file = $param_hash->{mutation_file};
+
+  my @matches;
+
+  open my $f, $mutation_file or die("ERROR: couldn't open file $mutation_file\n");
+  while(<$f>) {
+    my @fields = split /\t/;
+    my $ref = $fields[8];
+    my (undef, $des) = split /:/, $fields[1];
+    
+    next unless ($ref and $des);
+       
+    if( ($des eq $id_des) && ($ref eq $id_ref) ) {
+        push @matches, $_;
+    }
+  }
+
+  return \@matches;
+}
+
+# parse the extracted lines of IntAct data file
+sub _parse_intact_data {
+  my ($self, $data) = @_;
+
+  my $i = 0;
+  my %parsed_data = map { $valid_fields->{$i++} => $_ } split /\t/, $data
+    or die("ERROR: cannot parse intact data.\n");
+  
+  # Replace , and | to avoid confusion for interaction paritcipants
+  $parsed_data{interaction_participants} =~ s/\|/ and /g;
+  $parsed_data{interaction_participants} =~ s/,//g;
+   
+  return \%parsed_data;
+}
+
+# remove duplicate interactions
+sub _remove_duplicates {
+  my ($self, $intact_matches) = @_;
+  
+  my @interaction_acs;
+  my @feature_types;
+  my @interaction_participants;
+
+  my @matches;
+  foreach (@$intact_matches) {
+    my $parsed_data = $self->_parse_intact_data($_);
+
+    next if $parsed_data->{ap_organism} ne "9606 - Homo sapiens";
+    
+    my $interaction_ac = $parsed_data->{interaction_ac};
+    my $feature_type = $parsed_data->{feature_type};
+    my $interaction_participant = $parsed_data->{interaction_participants};
+    
+    if( grep { $interaction_ac } @interaction_acs &&
+        grep { $feature_type } @feature_types &&
+        grep { $interaction_participant } @interaction_participants) {
+
+      push @matches, $parsed_data;
+    
+      push @interaction_acs, $interaction_ac;
+      push @feature_types, $feature_type;
+      push @interaction_participants, $interaction_participant;
+    }
+  }
+  
+  return \@matches;
+}
+
+# filter out field according to user input
+sub _filter_fields {
+  my ($self, $uniq_matches) = @_;
+  
+  my (@arr, %hash);
+  
+  foreach (@$uniq_matches) {
+    # Add separator for default/tab output
+    $hash{"IntAct"} = $hash{"IntAct"}."|" if $hash{"IntAct"};
+
+    my $j = 0;
+    while (defined $valid_fields->{$j}) {
+      my $field = $valid_fields->{$j++};
+      my $field_val = $_->{$field};
+
+      chomp $field_val;
+
+      if(defined $self->{$field}) {
+        if($output_json || $output_rest){
+          $hash{$field} = $field_val;
+        }
+        elsif($output_vcf){
+          $hash{"IntAct_".$field} = $hash{"IntAct_".$field} ? $hash{"IntAct_".$field}.",".$field_val : $field_val;
+        }
+        else{
+          $hash{"IntAct"} = $hash{"IntAct"} ? $hash{"IntAct"}."<".$field_val.">" : "<".$field_val.">";
+        }
+      }
+    }
+
+    push @arr, \%hash if $output_json || $output_rest;
```

Comment:
This is pushing a reference to a hash in the array. The hash is over-written in the while loop, so we return multiple references to the same hash - the final record is reported multiple times.